### PR TITLE
SNOW-1791994 IR: Generate typed protobuf Python code

### PR DESCRIPTION
**Pull request by sfc-gh-azhan**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1791994

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

This PR generates typed Python code from protobuf and make sure CI check types in precommit. 

Copilot summary:
This pull request includes several changes to improve type checking, streamline the build process, and enhance code quality. The most important changes include the addition of `mypy-protobuf` for generating typed Python code from protobuf, updates to the build and pre-commit configurations, and improvements to type annotations across various files.

### Build and Dependency Management:
* Added `mypy-protobuf` installation in `.github/scripts/install_protoc.sh` and `scripts/jenkins_regress.sh` to generate typed Python code from protobuf. [[1]](diffhunk://#diff-0899f49e438b021e67e050dba42758e54dfb9f4004c38353a7b0824f691e50a0R74-R77) [[2]](diffhunk://#diff-b69c9f6a77a63d66300141897cfd4872f2f6c2ae6762f52c85b8824d973246ebL21-R21) [[3]](diffhunk://#diff-6b37b6a78b44ea8880f92d1263744ef1e0df235bf59f1af110b37f1068344498L18-R18)
* Updated `.github/workflows/precommit.yml` to install `tox` and streamline the documentation build process using `tox`.
* Modified `setup.py` to include `mypy-protobuf` as a dependency and updated the protobuf generation command to use `--mypy_out`. [[1]](diffhunk://#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R60) [[2]](diffhunk://#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7L84-R85) [[3]](diffhunk://#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7L118-R119)

### Pre-commit Configuration:
* Added new files and dependencies to the `mypy` check in `.pre-commit-config.yaml`. [[1]](diffhunk://#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R94) [[2]](diffhunk://#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R107-R113)

### Code Quality and Type Annotations:
* Enhanced type annotations in `src/snowflake/snowpark/_internal/ast/batch.py` by specifying return types and dictionary types. [[1]](diffhunk://#diff-8d22dcd969690348d0b396f6b73598ecd2e42e5ec20a503f518f444ec3c74311L67-R69) [[2]](diffhunk://#diff-8d22dcd969690348d0b396f6b73598ecd2e42e5ec20a503f518f444ec3c74311L87-R87) [[3]](diffhunk://#diff-8d22dcd969690348d0b396f6b73598ecd2e42e5ec20a503f518f444ec3c74311L105-R105)
* Improved type checking in `src/snowflake/snowpark/_internal/ast/utils.py` by adding `# type: ignore` comments and TODOs for better handling of type issues. [[1]](diffhunk://#diff-546ece6125c3622c5002e40620d5a4508bc40bd95e097f7261b3c82d361ec3ffL63-R63) [[2]](diffhunk://#diff-546ece6125c3622c5002e40620d5a4508bc40bd95e097f7261b3c82d361ec3ffL83-R90) [[3]](diffhunk://#diff-546ece6125c3622c5002e40620d5a4508bc40bd95e097f7261b3c82d361ec3ffL139-R142) [[4]](diffhunk://#diff-546ece6125c3622c5002e40620d5a4508bc40bd95e097f7261b3c82d361ec3ffL151-R165) [[5]](diffhunk://#diff-546ece6125c3622c5002e40620d5a4508bc40bd95e097f7261b3c82d361ec3ffL185-R185) [[6]](diffhunk://#diff-546ece6125c3622c5002e40620d5a4508bc40bd95e097f7261b3c82d361ec3ffL194-R231)

Tested Jenkins jobs:
- https://ci-dev-142.int.snowflakecomputing.com/job/SnowparkPythonSnowflakePythonClientRegressRunner/799/console
- https://ci-dev-142.int.snowflakecomputing.com/job/SnowparkPandasStoredProcPrecommitTest/193/console

### CODE REVIEW

File: .github/workflows/precommit.yml [hunk 1]
```diff
@@ -593,19 +593,13 @@ jobs:
...
-        run: python -m pip install ".[modin-development]"
-      - name: Install Sphinx
-        run: python -m pip install sphinx
       - name: Build document
-        working-directory: docs
-        # treat warning as failure but complete the entire process
-        run: |
-          make clean
-          make html SPHINXOPTS="-W --keep-going"
+        run: python -m tox -e docs
```
Comment (sfc-gh-azhan) [inline .github/workflows/precommit.yml:602]: just reuse the same code in tox

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -91,6 +91,7 @@ repos:
       - id: mypy
         files: >
           (?x)^(
+            src/snowflake/snowpark/_internal/ast/.*\.py|
```
Comment (sfc-gh-azhan) [inline .pre-commit-config.yaml:94]: cover all ast code in mypy type check

File: src/snowflake/snowpark/_internal/ast/utils.py [hunk 3]
```diff
@@ -80,14 +80,14 @@ def __init__(self) -> None:  # pragma: no cover
         self.symbols: Optional[Union[str, List[str]]] = None
 
     # TODO(SNOW-1491199) - This method is not covered by tests until the end of phase 0. Drop the pragma when it is covered.
-    def visit_Assign(self, node) -> None:  # pragma: no cover
+    def visit_Assign(self, node: ast.Assign) -> None:  # pragma: no cover
```
Comment (sfc-gh-azhan) [inline src/snowflake/snowpark/_internal/ast/utils.py:83]: we should fix those once our code is in.

File: src/snowflake/snowpark/_internal/ast/utils.py [hunk 4]
```diff
@@ -470,17 +470,17 @@ def build_indirect_table_fn_apply(
...
+    expr = with_src_position(ast.apply_expr)  # type: ignore[arg-type] # TODO(SNOW-1491199) # Argument 1 to "with_src_position" has incompatible type "ApplyExpr"; expected "Expr"
     if isinstance(
-        func, (snowflake.snowpark.table_function.TableFunctionCall, Callable)
+        func, (snowflake.snowpark.table_function.TableFunctionCall, Callable)  # type: ignore[arg-type] # TODO(SNOW-1491199) # Argument 2 to "isinstance" has incompatible type "tuple[type[TableFunctionCall], <typing special form>]"; expected "_ClassInfo"
     ):
-        stmt = func._ast_stmt
-        fn_expr = expr.fn.indirect_table_fn_id_ref
+        stmt = func._ast_stmt  # type: ignore[union-attr] # TODO(SNOW-1491199) # Item "str" of "Union[str, list[str], TableFunctionCall, Callable[..., Any]]" has no attribute "_ast_stmt", Item "list[str]" of "Union[str, list[str], TableFunctionCall, Callable[..., Any]]" has no attribute "_ast_stmt", Item "TableFunctionCall" of "Union[str, list[str], TableFunctionCall, Callable[..., Any]]" has no attribute "_ast_stmt", Item "function" of "Union[str, list[str], TableFunctionCall, Callable[..., Any]]" has no attribute "_ast_stmt"
```
Comment (sfc-gh-oplaton) [inline src/snowflake/snowpark/_internal/ast/utils.py:477]: I assume the comments are machine-generated. Does any particular linter or annotation checker require the full message? Looks pretty cumbersome.

**Comment by sfc-gh-oplaton** [comment]
Thank you for setting this up! Generally looks good. I did not review `_internal/ast/*.py` in detail, I assume all comments are machine-generated to silence warnings.

File: setup.py [hunk 5]
```diff
@@ -57,6 +57,7 @@
     "pytest-assume",  # sql counter check
     "decorator",  # sql counter check
     "protoc-wheel-0==21.1",  # Protocol buffer compiler, for Snowpark IR
+    "mypy-protobuf",  # used in generating typed Python code from protobuf for Snowpark IR
```
Comment (sfc-gh-oplaton) [inline setup.py:60]: I assume this has no licensing or security concerns, either. Should this require a specific minimum version?

**Review by sfc-gh-oplaton — APPROVED** [review]

File: setup.py (hunk 5 above)
Comment (sfc-gh-azhan) [reply setup.py:60]: Yes, this is development requirements so should be fine.

File: src/snowflake/snowpark/_internal/ast/utils.py (hunk 4 above)
Comment (sfc-gh-azhan) [reply src/snowflake/snowpark/_internal/ast/utils.py:477]: Yes, it contains three parts:
1. jira no. to track
2. the type ignore 
3. the error message
It's probably verbose for the error message but I try to capture them so when we fix them later, we can easily know what was the issue before.

**Review by sfc-gh-aalam — APPROVED** [review]

**Pull request merged**